File: bindings/nodejs/axiomdb.mjs
/**
 * AxiomDB Node.js binding — koffi FFI over libaxiomdb_embedded.
 *
 * Result materialization uses the packed-buffer path: one FFI call returns the
 * whole result as a single contiguous buffer, parsed once in JS. This avoids
 * the ~2 FFI crossings per cell of the naive accessor loop (which was ~9×
 * slower than better-sqlite3); the packed path is ~2× — a 4× improvement.
 *
 * For full parity with better-sqlite3 (a native C++ addon), see the native
 * N-API addon in `bindings/axiomdb-node` (built with napi-rs).
 *
 * Usage:
 *   import { AxiomDB } from './axiomdb.mjs';
 *   const db = new AxiomDB('./myapp.db');
 *   db.execute("CREATE TABLE users (id INT PRIMARY KEY, name TEXT)");
 *   db.execute("INSERT INTO users VALUES (1, 'Alice')");
 *   db.query("SELECT * FROM users");        // [{ id: 1, name: 'Alice' }]
 *   db.queryTuples("SELECT * FROM users");  // [[1, 'Alice']]  (fastest)
 *   db.close();
 */

import koffi from 'koffi';
import { existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

// ── Library loading ─────────────────────────────────────────────────────────

const __dirname = dirname(fileURLToPath(import.meta.url));

function findLibrary() {
  const platform = process.platform;
  const ext = platform === 'darwin' ? 'dylib' : platform === 'win32' ? 'dll' : 'so';
  const name = `libaxiomdb_embedded.${ext}`;
  const paths = [
    resolve(__dirname, name),
    resolve(__dirname, '..', '..', 'target', 'release', name),
    resolve(__dirname, '..', '..', 'target', 'debug', name),
    `/usr/local/lib/${name}`,
  ];
  for (const p of paths) {
    if (existsSync(p)) return p;
  }
  return name;
}

const lib = koffi.load(findLibrary());

// ── Type constants (packed cell tags) ────────────────────────────────────────

const TYPE_NULL = 0;
const TYPE_INTEGER = 1;
const TYPE_REAL = 2;
const TYPE_TEXT = 3;
const TYPE_BLOB = 4;
const PACKED_MAGIC = 0x41584d31; // "AXM1"

// ── C function declarations ───────────────────────────────────────────────────

const axiomdb_open = lib.func('void* axiomdb_open(const char* path)');
const axiomdb_execute = lib.func('int64_t axiomdb_execute(void* db, const char* sql)');
const axiomdb_close = lib.func('void axiomdb_close(void* db)');
const axiomdb_last_error = lib.func('const char* axiomdb_last_error(void* db)');
// Packed result buffer: one FFI call returns the whole result set.
const axiomdb_query_packed = lib.func(
  'uint8_t* axiomdb_query_packed(void* db, const char* sql, _Out_ size_t* out_len)'
);
const axiomdb_packed_free = lib.func('void axiomdb_packed_free(uint8_t* ptr, size_t len)');

// ── Packed buffer parser ──────────────────────────────────────────────────────

/**
 * Parses a packed result buffer into { columns, rows } where each row is an
 * array of values. Format (LE): magic, n_cols, n_rows, column names, then
 * per-cell tag + payload.
 */
function parsePacked(buf) {
  let off = 0;
  const magic = buf.readUInt32LE(off); off += 4;
  if (magic !== PACKED_MAGIC) {
    throw new Error(`corrupt packed buffer (magic=0x${magic.toString(16)})`);
  }
  const nCols = buf.readUInt32LE(off); off += 4;
  const nRows = Number(buf.readBigUInt64LE(off)); off += 8;

  const columns = new Array(nCols);
  for (let c = 0; c < nCols; c++) {
    const len = buf.readUInt32LE(off); off += 4;
    columns[c] = buf.toString('utf8', off, off + len); off += len;
  }

  const rows = new Array(nRows);
  for (let r = 0; r < nRows; r++) {
    const row = new Array(nCols);
    for (let c = 0; c < nCols; c++) {
      const tag = buf[off++];
      if (tag === TYPE_INTEGER) {
        row[c] = Number(buf.readBigInt64LE(off)); off += 8;
      } else if (tag === TYPE_TEXT) {
        const len = buf.readUInt32LE(off); off += 4;
        row[c] = buf.toString('utf8', off, off + len); off += len;
      } else if (tag === TYPE_REAL) {
        row[c] = buf.readDoubleLE(off); off += 8;
      } else if (tag === TYPE_BLOB) {
        const len = buf.readUInt32LE(off); off += 4;
        row[c] = Buffer.from(buf.subarray(off, off + len)); off += len; // owned copy
      } else {
        row[c] = null; // TYPE_NULL
      }
    }
    rows[r] = row;
  }
  return { columns, rows };
}

// ── JavaScript API ──────────────────────────────────────────────────────────

export class AxiomDB {
  #ptr = null;

  /** Open or create a database at the given file path. */
  constructor(path) {
    this.#ptr = axiomdb_open(path);
    if (!this.#ptr) {
      throw new Error(`Failed to open database at '${path}'`);
    }
  }

  /** Execute a DDL/DML statement. Returns rows affected. */
  execute(sql) {
    this.#checkOpen();
    const result = axiomdb_execute(this.#ptr, sql);
    if (result < 0) {
      throw new Error(this.#lastError() || 'execute failed');
    }
    return Number(result);
  }

  /** Run a SELECT via the packed buffer (one FFI call). Internal. */
  #queryPacked(sql) {
    this.#checkOpen();
    const lenOut = [0n];
    const ptr = axiomdb_query_packed(this.#ptr, sql, lenOut);
    if (!ptr) {
      throw new Error(this.#lastError() || 'query failed');
    }
    const len = Number(lenOut[0]);
    try {
      // `koffi.view` is a Buffer-compatible window into native memory (not an
      // owned copy). parsePacked must run BEFORE the buffer is freed; it copies
      // every value into owned JS objects (strings/numbers/blob Buffers), so the
      // returned result stays valid afterwards.
      const buf = Buffer.from(koffi.view(ptr, len));
      return parsePacked(buf);
    } finally {
      axiomdb_packed_free(ptr, len);
    }
  }

  /**
   * Execute a SELECT and return rows as an array of objects
   * (column name → value).
   */
  query(sql) {
    const { columns, rows } = this.#queryPacked(sql);
    return rows.map((row) => {
      const obj = {};
      for (let c = 0; c < columns.length; c++) obj[columns[c]] = row[c];
      return obj;
    });
  }

  /**
   * Execute a SELECT and return rows as arrays (fastest path; matches
   * better-sqlite3's `.raw().all()` shape).
   */
  queryTuples(sql) {
    return this.#queryPacked(sql).rows;
  }

  /** Execute a SELECT and return { columns, rows } (rows as arrays). */
  queryWithColumns(sql) {
    return this.#queryPacked(sql);
  }

  /** Close the database. Safe to call multiple times. */
  close() {
    if (this.#ptr) {
      axiomdb_close(this.#ptr);
      this.#ptr = null;
    }
  }

  /** Return the last error message, or null. */
  lastError() {
    return this.#lastError();
  }

  #checkOpen() {
    if (!this.#ptr) throw new Error('Database is closed');
  }

  #lastError() {
    if (!this.#ptr) return null;
    return axiomdb_last_error(this.#ptr) || null;
  }
}

File: bindings/swift/Sources/AxiomDB/AxiomDB.swift
import CAxiomDB
import Foundation

/// A single result cell.
public enum AxiomValue: Equatable, Sendable {
    case null
    case int(Int64)
    case double(Double)
    case text(String)
    case blob([UInt8])
}

public struct AxiomError: Error, CustomStringConvertible {
    public let message: String
    public var description: String { "AxiomDB: \(message)" }
}

/// An in-process AxiomDB database.
///
/// Materialization uses the packed-buffer path: one FFI call returns the whole
/// result as a single buffer, parsed once in Swift. Swift's C interop is direct
/// (no marshalling), and the parser uses `loadUnaligned`, so this is compiled,
/// allocation-light decoding.
///
/// Not safe for concurrent use from multiple threads; use one instance per
/// thread (like `sqlite3` with `check_same_thread`).
public final class AxiomDB {
    private var ptr: OpaquePointer?

    private static let packedMagic: UInt32 = 0x4158_4D31  // "AXM1"

    /// Opens or creates a database at `path` (`":memory:"` for ephemeral).
    public init(_ path: String) throws {
        ptr = axiomdb_open(path)
        if ptr == nil { throw AxiomError(message: "failed to open database at \(path)") }
    }

    deinit { close() }

    /// Closes the database. Safe to call multiple times.
    public func close() {
        if let p = ptr {
            axiomdb_close(p)
            ptr = nil
        }
    }

    /// Returns the last error message, or `nil`.
    public func lastError() -> String? {
        guard let p = ptr, let msg = axiomdb_last_error(p) else { return nil }
        return String(cString: msg)
    }

    /// Executes a DDL/DML statement. Returns rows affected.
    ///
    /// Pass `params` to bind `?` placeholders with real prepared-statement
    /// binding — no string interpolation, no SQL injection:
    ///
    /// ```swift
    /// try db.execute("INSERT INTO t VALUES (?, ?)", [.int(1), .text("Alice")])
    /// ```
    @discardableResult
    public func execute(_ sql: String, _ params: [AxiomValue] = []) throws -> Int64 {
        guard let p = ptr else { throw AxiomError(message: "database is closed") }
        let n: Int64
        if params.isEmpty {
            n = axiomdb_execute(p, sql)
        } else {
            let enc = Self.encodeParams(params)
            n = enc.withUnsafeBufferPointer { bp in
                axiomdb_execute_params(p, sql, bp.baseAddress, enc.count)
            }
        }
        if n < 0 { throw AxiomError(message: lastError() ?? "execute failed") }
        return n
    }

    /// Executes a SELECT, returning rows as arrays (fastest).
    /// Pass `params` to bind `?` placeholders safely (see `execute`).
    public func queryTuples(_ sql: String, _ params: [AxiomValue] = []) throws -> [[AxiomValue]] {
        try queryPacked(sql, params).rows
    }

    /// Executes a SELECT, returning column names plus rows.
    /// Pass `params` to bind `?` placeholders safely (see `execute`).
    public func queryWithColumns(_ sql: String, _ params: [AxiomValue] = []) throws -> (
        columns: [String], rows: [[AxiomValue]]
    ) {
        try queryPacked(sql, params)
    }

    /// Executes a SELECT, returning rows as dictionaries (column name → value).
    /// Pass `params` to bind `?` placeholders safely (see `execute`).
    public func query(_ sql: String, _ params: [AxiomValue] = []) throws -> [[String: AxiomValue]] {
        let (cols, rows) = try queryPacked(sql, params)
        return rows.map { row in
            var dict = [String: AxiomValue](minimumCapacity: cols.count)
            for (i, name) in cols.enumerated() { dict[name] = row[i] }
            return dict
        }
    }

    @discardableResult public func begin() throws -> Int64 { try execute("BEGIN") }
    @discardableResult public func commit() throws -> Int64 { try execute("COMMIT") }
    @discardableResult public func rollback() throws -> Int64 { try execute("ROLLBACK") }

    // MARK: - Internal

    private func queryPacked(_ sql: String, _ params: [AxiomValue] = []) throws -> (
        columns: [String], rows: [[AxiomValue]]
    ) {
        guard let p = ptr else { throw AxiomError(message: "database is closed") }
        var len: size_t = 0
        let result: UnsafeMutablePointer<UInt8>?
        if params.isEmpty {
            result = axiomdb_query_packed(p, sql, &len)
        } else {
            let enc = Self.encodeParams(params)
            result = enc.withUnsafeBufferPointer { bp in
                axiomdb_query_packed_params(p, sql, bp.baseAddress, enc.count, &len)
            }
        }
        guard let buf = result else {
            throw AxiomError(message: lastError() ?? "query failed")
        }
        defer { axiomdb_packed_free(buf, len) }  // freed after parse
        return Self.parsePacked(UnsafeRawPointer(buf), Int(len))
    }

    /// Serializes positional `params` into the AxiomDB param buffer: `u32 count`,
    /// then per-param `{u8 tag, payload}`. Tags mirror the packed cell encoding:
    /// 0=null, 1=int(i64), 2=double(f64), 3=text, 4=blob.
    static func encodeParams(_ params: [AxiomValue]) -> [UInt8] {
        var buf = [UInt8]()
        buf.reserveCapacity(4 + params.count * 9)
        func putU32(_ v: UInt32) { withUnsafeBytes(of: v.littleEndian) { buf.append(contentsOf: $0) } }
        func putI64(_ v: Int64) { withUnsafeBytes(of: v.littleEndian) { buf.append(contentsOf: $0) } }
        func putF64(_ v: Double) {
            withUnsafeBytes(of: v.bitPattern.littleEndian) { buf.append(contentsOf: $0) }
        }
        putU32(UInt32(params.count))
        for p in params {
            switch p {
            case .null:
                buf.append(0)
            case .int(let i):
                buf.append(1)
                putI64(i)
            case .double(let d):
                buf.append(2)
                putF64(d)
            case .text(let s):
                buf.append(3)
                let bytes = Array(s.utf8)
                putU32(UInt32(bytes.count))
                buf.append(contentsOf: bytes)
            case .blob(let b):
                buf.append(4)
                putU32(UInt32(b.count))
                buf.append(contentsOf: b)
            }
        }
        return buf
    }

    /// Decodes a packed (AXM1) buffer. Reads via `loadUnaligned` (native LE on
    /// arm64/x86); strings via `String(decoding:as:)`. Every value is copied
    /// into Swift storage, so the result is valid after the buffer is freed.
    static func parsePacked(_ base: UnsafeRawPointer, _ len: Int) -> (columns: [String], rows: [[AxiomValue]]) {
        var off = 0
        func u32() -> Int { let v = base.loadUnaligned(fromByteOffset: off, as: UInt32.self); off += 4; return Int(v) }
        func u64() -> Int { let v = base.loadUnaligned(fromByteOffset: off, as: UInt64.self); off += 8; return Int(v) }
        func i64() -> Int64 { let v = base.loadUnaligned(fromByteOffset: off, as: Int64.self); off += 8; return v }
        func f64() -> Double { let v = base.loadUnaligned(fromByteOffset: off, as: Double.self); off += 8; return v }
        func text(_ n: Int) -> String {
            let s = String(decoding: UnsafeRawBufferPointer(start: base + off, count: n), as: UTF8.self)
            off += n
            return s
        }
        func blob(_ n: Int) -> [UInt8] {
            let arr = [UInt8](UnsafeRawBufferPointer(start: base + off, count: n))
            off += n
            return arr
        }

        let magic = base.loadUnaligned(fromByteOffset: 0, as: UInt32.self)
        off = 4
        precondition(magic == packedMagic, "corrupt packed buffer")
        let nCols = u32()
        let nRows = u64()

        var columns = [String]()
        columns.reserveCapacity(nCols)
        for _ in 0..<nCols { let l = u32(); columns.append(text(l)) }

        var rows = [[AxiomValue]]()
        rows.reserveCapacity(nRows)
        for _ in 0..<nRows {
            var row = [AxiomValue]()
            row.reserveCapacity(nCols)
            for _ in 0..<nCols {
                let tag = base.load(fromByteOffset: off, as: UInt8.self)
                off += 1
                switch tag {
                case 1: row.append(.int(i64()))
                case 3: let l = u32(); row.append(.text(text(l)))
                case 2: row.append(.double(f64()))
                case 4: let l = u32(); row.append(.blob(blob(l)))
                default: row.append(.null)
                }
            }
            rows.append(row)
        }
        return (columns, rows)
    }
}
